### PR TITLE
Verify strictly positive for gamma regression.

### PR DESCRIPTION
**Pull request by trivialfis**
Close https://github.com/dmlc/xgboost/issues/6723 .

### CODE REVIEW

File: src/metric/elementwise_metric.cu [hunk 1]
```diff
@@ -304,13 +304,17 @@ struct EvalGammaNLogLik {
   }
 
   XGBOOST_DEVICE bst_float EvalRow(bst_float y, bst_float py) const {
-    const bst_float eps = 1e-16f;
-    if (y < eps) y = eps;
-    bst_float psi = 1.0;
+    py = std::max(py, 1e-6f);
+    // hardcoded dispersion.
+    float constexpr kPsi = 1.0;
```
Comment (hcho3) [inline src/metric/elementwise_metric.cu:309]: Do we have a plan to allow for a different value for dispersion in the future?

File: src/metric/elementwise_metric.cu [hunk 2]
```diff
@@ -304,13 +304,17 @@ struct EvalGammaNLogLik {
...
-    bst_float a = psi;
-    bst_float b = -std::log(-theta);
-    bst_float c = 1. / psi * std::log(y/psi) - std::log(y) - common::LogGamma(1. / psi);
+    bst_float a = kPsi;
+    // b = -std::log(-theta);
+    float b = 1.0f;
+    // c = 1. / kPsi * std::log(y/kPsi) - std::log(y) - common::LogGamma(1. / kPsi);
+    //     1.0f      * 1.0f             - 1.0f        - 0
+    float c = 0;
+    // general form for exponential family.
```
Comment (hcho3) [inline src/metric/elementwise_metric.cu:317]: Did you manage to locate a reference for this?

File: src/metric/elementwise_metric.cu (hunk 1 above)
Comment (trivialfis) [reply src/metric/elementwise_metric.cu:309]: The dispersion should be estimated from data.  We can put more efforts to clarity the choice or add estimation in the code,right now I'm just adding notes.

Comment (hcho3) [reply src/metric/elementwise_metric.cu:309]: It can be treated as a hyperparameter.

Comment (trivialfis) [reply src/metric/elementwise_metric.cu:309]: I'm not sure yet,as you can see the metric doesn't use label at all.

Comment (trivialfis) [reply src/metric/elementwise_metric.cu:309]: The dependency on labels is hidden in dispersion.

Comment (hcho3) [reply src/metric/elementwise_metric.cu:309]: Wait, why is `std::log(y)` equal to 1.0f?

File: src/metric/elementwise_metric.cu [hunk 3]
```diff
@@ -304,13 +304,17 @@ struct EvalGammaNLogLik {
...
+    float constexpr kPsi = 1.0;
     bst_float theta = -1. / py;
-    bst_float a = psi;
-    bst_float b = -std::log(-theta);
-    bst_float c = 1. / psi * std::log(y/psi) - std::log(y) - common::LogGamma(1. / psi);
+    bst_float a = kPsi;
+    // b = -std::log(-theta);
+    float b = 1.0f;
+    // c = 1. / kPsi * std::log(y/kPsi) - std::log(y) - common::LogGamma(1. / kPsi);
+    //     1.0f      * 1.0f             - 1.0f        - 0
```
Comment (hcho3) [inline src/metric/elementwise_metric.cu:315]: If `y` is a positive value, `std::log(y)` is not necessarily equal to 1.0f, I think.

File: src/metric/elementwise_metric.cu (hunk 2 above)
Comment (trivialfis) [reply src/metric/elementwise_metric.cu:317]: All exponential distribution can be expressed in this form disregarding parametrisation.

File: src/metric/elementwise_metric.cu (hunk 3 above)
Comment (trivialfis) [reply src/metric/elementwise_metric.cu:315]: There are 2 log y, they cancel out.

File: src/metric/elementwise_metric.cu (hunk 1 above)
Comment (trivialfis) [reply src/metric/elementwise_metric.cu:309]: Where?

File: src/metric/elementwise_metric.cu (hunk 3 above)
Comment (hcho3) [reply src/metric/elementwise_metric.cu:315]: Thanks, I will suggest a small change to clarify this point.

File: src/metric/elementwise_metric.cu (hunk 1 above)
Comment (hcho3) [reply src/metric/elementwise_metric.cu:309]: See my next comment, which you just responded.

File: src/metric/elementwise_metric.cu (hunk 3 above)
Comment (trivialfis) [reply src/metric/elementwise_metric.cu:315]: The 1.0 is misleading, will remove it.

Comment (hcho3) [inline src/metric/elementwise_metric.cu:315]:
```suggestion
    // c = 1. / kPsi * std::log(y/kPsi) - std::log(y) - common::LogGamma(1. / kPsi);
    //   = 1.0f      * std::log(y)      - std::log(y) - 0 = 0
```

**Review by hcho3 — APPROVED** [review]

**Review by RAMitchell — APPROVED** [review]
